### PR TITLE
Fix bug in CPUManager with race on container map access

### PR DESCRIPTION
**Pull request by klueska**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Fixes a sporadic bug that causes the `kubelet` to crash:
```
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: fatal error: concurrent map iteration and map write
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: goroutine 141 [running]:
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: runtime.throw(0x43a5919, 0x26)
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]:         /usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc0036eaca8 sp=0xc0036eac78 pc=0x432eb2
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: runtime.mapiternext(0xc0036eadb0)
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]:         /usr/local/go/src/runtime/map.go:858 +0x579 fp=0xc0036ead30 sp=0xc0036eaca8 pc=0x412db9
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/containermap.ContainerMap.GetContainerID(0xc000cae360, 0xc002671350, 0x24, 0xc0038c15a0, 0xb, 0xc0038c15a0, 0xb, 0x0, 0x0)
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/containermap/container_map.go:57 +0x9e fp=0xc0036eae20 sp=0xc0036ead30 pc=0x1b2806e
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/containermap.ContainerMap.RemoveByContainerRef(...)
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]:         /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/cm/cpumanager/containermap/container_map.go:49
Dec 19 12:01:13 ngcdgx2k8s0094d kubelet[69329]: k8s.io/kubernetes/pkg/kubelet/cm/cpumanager.(*manager).policyRemoveContainerByRef(0xc00065c990, 0xc002671350, 0x24, 0xc0038c15a0, 0xb, 0x2, 0x2)
```

```release-note
Fixed bug in CPUManager with race on container map access
```

### CODE REVIEW

File: pkg/kubelet/cm/cpumanager/cpu_manager.go
Comment (andrewsykim) [inline pkg/kubelet/cm/cpumanager/cpu_manager.go:405]: nit: I think this would read slightly easier to follow if we called `m.Lock()` twice, each time before invoking the methods accessing `containerMap`. Just easier to track where and why the lock is needed.

**Comment by odinuge** [comment]
Change looks good to me, but agree with comment from @andrewsykim, so i'll let @klueska look at that.

/lgtm
/hold

File: pkg/kubelet/cm/cpumanager/cpu_manager.go
Comment (klueska) [reply pkg/kubelet/cm/cpumanager/cpu_manager.go:405]: I'm actually happy with the way it is currently (I've been stress testing it like this in a soak environment over the last 2 weeks without issue). If you feel strongly though, I can add small helper functions to do the locking around each call.

**Comment by klueska** [comment]
I chatted with @andrewsykim offline and, while I tend to agree with him on some level, I also think it’s pretty clear as is (and I’ve been stress testing it in its current form for 2 weeks without issue). He has agreed that his issue is non-blocking and we've agreed to merge this as-is.

/unhold

**Review by hasheddan — APPROVED** [review]
/lgtm

@klueska I'll wait until this one merges to go through cherry-picks :+1: Feel free to ping me if I don't follow up promptly.

**Pull request merged**
